File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& heig) {
        int l = heig[0];
        int r = heig[heig.size()-1];

        int x=0;
        int y=heig.size()-1;
        int ans=0,p;
        while (x<=y){
            if(l<=r){
                l=max(l,heig[x]);
                ans+=max(0,l-heig[x]);
                x+=1;
            }else{
                r=max(r,heig[y]);
                ans+=max(0,r-heig[y]);
                y-=1;
            }

        }
        return ans;
        
    }
};

File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:

        leftmax = height[0]
        rightmax = height[-1]


        left = 0
        right = len(height)-1
        ans = 0
        while left <= right:
            if leftmax<rightmax:
                ans += max(0,leftmax -height[left])
                leftmax= max(leftmax,height[left])
                left+=1
            else:
                ans += max(0,rightmax - height[right])
                rightmax= max(rightmax,height[right])
                right-=1
            print(ans,left,right)
        return ans
